compiler/noirc_frontend/src/hir_def/iterative_unification.rs

```diff
@@ -0,0 +1,339 @@
+                                bindings,
+                                |bindings| {
+                                    let only_integer = matches!(typ, Type::Integer(..));
+                                    other.try_bind_to_polymorphic_int(var, bindings, only_integer)
+                                },
+                            )
+                        } else {
+                            self.try_unify_to_type_variable_iter(
+                                other_id,
+                                var,
+                                bindings,
+                                |bindings| other.try_bind_to(var, bindings, typ.kind()),
+                            )
+                        }
+                    }
+                    TypeBinding::Unbound(_id, Kind::IntegerOrField) => self
+                        .try_unify_to_type_variable_iter(other_id, var, bindings, |bindings| {
+                            let only_integer = false;
+                            other.try_bind_to_polymorphic_int(var, bindings, only_integer)
+                        }),
+                    TypeBinding::Unbound(_id, Kind::Integer) => self
+                        .try_unify_to_type_variable_iter(other_id, var, bindings, |bindings| {
+                            let only_integer = true;
+                            other.try_bind_to_polymorphic_int(var, bindings, only_integer)
+                        }),
+                    TypeBinding::Unbound(_id, type_var_kind) => self
+                        .try_unify_to_type_variable_iter(other_id, var, bindings, |bindings| {
+                            other.try_bind_to(var, bindings, type_var_kind.clone())
+                        }),
+                }
+            }
+
+            (Array(len_a, elem_a), Array(len_b, elem_b)) => {
+                Ok(vec![self.for_unite(len_a, len_b), self.for_unite(elem_a, elem_b)])
+            }
+
+            (Slice(elem_a), Slice(elem_b)) => Ok(vec![self.for_unite(elem_a, elem_b)]),
+
+            (String(len_a), String(len_b)) => Ok(vec![self.for_unite(len_a, len_b)]),
+
+            (FmtString(len_a, elements_a), FmtString(len_b, elements_b)) => {
+                Ok(vec![self.for_unite(len_a, len_b), self.for_unite(elements_a, elements_b)])
+            }
+
+            (Tuple(elements_a), Tuple(elements_b)) => {
+                if elements_a.len() != elements_b.len() {
+                    Err(UnificationError)
+                } else {
+                    let mut to_unit = Vec::new();
+                    for (a, b) in elements_a.iter().zip(elements_b) {
+                        to_unit.push(self.for_unite(a, b));
+                    }
+                    Ok(to_unit)
+                }
+            }
+
+            // No recursive try_unify call for struct fields. Don't want
+            // to mutate shared type variables within struct definitions.
+            // This isn't possible currently but will be once noir gets generic types
```

Comment:
This looks like an outdated comment cc @jfecher

Reply:
Done